test/relayclient/RelayClient.test.ts

```diff
@@ -156,6 +156,44 @@ contract('RelayClient', function (accounts) {
       assert.equal(relayingErrors.size, 1)
       assert.equal(relayingErrors.get(localhostOne)!.message, BadHttpClient.message)
     })
+
+    it('should return errors in callback (asyncApprovalData) ', async function () {
+      const relayClient =
+        new RelayClient(underlyingProvider, gsnConfig, {
+          asyncApprovalData: async () => { throw new Error('approval-error') }
+        })
+      const { transaction, relayingErrors, pingErrors } = await relayClient.relayTransaction(options)
+      assert.isUndefined(transaction)
+      assert.equal(pingErrors.size, 0)
+      assert.equal(relayingErrors.size, 1)
+      assert.match(relayingErrors.values().next().value.message, /approval-error/)
+    })
+
+    it('should return errors in callback (asyncPaymasterData) ', async function () {
+      const relayClient =
+        new RelayClient(underlyingProvider, gsnConfig, {
+          asyncPaymasterData: async () => { throw new Error('paymasterData-error') }
+        })
+      const { transaction, relayingErrors, pingErrors } = await relayClient.relayTransaction(options)
+      assert.isUndefined(transaction)
+      assert.equal(pingErrors.size, 0)
+      assert.equal(relayingErrors.size, 1)
+      assert.match(relayingErrors.values().next().value.message, /paymasterData-error/)
+    })
+
+    it.skip('should return errors in callback (scoreCalculator) ', async function () {
```

Comment:
is this test added skipped on purpose?

Reply:
yes: I added it for completeness, but it fails, since unused function doesn't throw....
we forgot to propagate the scoring function to the RelaysManager
(its a separate issue, and I think we also need to modify the scoring function: it should have a "next" param, so you can "chain" the current one with extra logic, otherwise you need to re-implment it entirely from scratch (e.g. handling failures)